sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -367,15 +368,16 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         throw new AnalysisException(s"Table $tableIdent already exists.")
 
       case _ =>
-        val cmd =
-          CreateTableUsingAsSelect(
-            tableIdent,
-            source,
-            partitioningColumns.map(_.toArray).getOrElse(Array.empty[String]),
-            getBucketSpec,
-            mode,
-            extraOptions.toMap,
-            df.logicalPlan)
+        val tableDesc = CatalogTable(
+          identifier = tableIdent,
+          tableType = CatalogTableType.EXTERNAL,
+          storage = CatalogStorageFormat.empty.copy(properties = extraOptions.toMap),
+          schema = new StructType,
+          provider = Some(source),
+          partitionColumnNames = partitioningColumns.getOrElse(Nil),
+          bucketSpec = getBucketSpec
+        )
+        val cmd = CreateTable(tableDesc, mode, Some(df.logicalPlan))
```

Comment:
hmmm, do we have to use `Option` even though the parameter is guaranteed to be not null?

For this case, we can't use `Option`, or the behaviour will be changed. Previously if `df.logicalPlan` is null, it's a bug and we will throw NPE somewhere. If we use `Option` here, then we are silently converting a CTAS to CREATE TABLE, which is not expected.